### ISSUE TITLE
'Per example' net output 

### ISSUE THREAD

**Issue by TalFurman** [apache/incubator-mxnet#6300]
Hi,

MXnet newbie.
I'm running a 'Lenet' like arch. with 'softmax_cross_entropy' loss, in a 'step-by-step' style (i.e not mod.fit...
instead using 'bind','init','forward' etc).
I'm trying to get a 'per-example' loss value for the examples in my batch.
How can this be done with mxnet?

Thanks in advance,

Tal 

"""""""""""""" code

data = mx.symbol.Variable('data')
lbl = mx.sym.Variable('softmax_label')

conv1 = mx.sym.Convolution(data=data, kernel=(5, 5), num_filter=20)
relu1 = mx.sym.Activation(data=conv1, act_type="relu")
pool1 = mx.sym.Pooling(data=relu1, pool_type="max", kernel=(2, 2), stride=(2, 2))

conv2 = mx.sym.Convolution(data=pool1, kernel=(5, 5), num_filter=10)
relu2 = mx.sym.Activation(data=conv2, act_type="relu")
pool2 = mx.sym.Pooling(data=relu1, pool_type="max", kernel=(2, 2), stride=(2, 2))

# first fullc layer
flatten = mx.sym.Flatten(data=pool2)
fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=500)
relu3 = mx.sym.Activation(data=fc1, act_type="relu")
# second fullc layer
fc2 = mx.sym.FullyConnected(data=relu3, num_hidden=1)
# softmax loss
lenet = mx.sym.SoftmaxOutput(data=fc2, name='softmax')
lenet = mx.sym.softmax_cross_entropy(fc2, lbl)

# initial state
mod = mx.mod.Module(symbol=lenet)

# bind, tell the module the data and label shapes, so
# that memory could be allocated on the devices for computation

mod.bind(data_shapes=train_iter.provide_data, label_shapes=train_iter.provide_label)

# init parameters
mod.init_params(initializer=mx.init.Xavier(magnitude=2.))

# init optimizer
mod.init_optimizer(optimizer='sgd', optimizer_params=(('learning_rate', 0.1), ))

####################################### Training
# use accuracy as the metric

# Data
data=train_img
part_data_size = 60000
part_data=data[:part_data_size]
part_label=train_lbl[:part_data_size]
num_samp=part_data.shape[0]
num_batch = np.ceil(num_samp/batch_size)

# Metric
metric = mx.metric.create('acc')

# Tree params
newPlayer=pPlayer(num_samp,1)
tree_lrn_rate=10**-4
tree_samp_prob=0.5
bDebug=False
batch_size = 2
Eps = 10**-5
with mx.Context(mx.gpu(1)):

    for epoch in np.arange(10):
        for batch_num in np.arange(num_batch):
            ind_vec=[]
            prob_vec=[]
            for sampleInd in np.arange(batch_size):
                [ind, prob] = newPlayer.sample(tree_samp_prob)
                ind_vec.append(ind)
                prob_vec.append(prob)
            this_batch_img=train_img[ind_vec]
            this_batch_lbl=train_lbl[ind_vec]

            #train_iter = mx.io.NDArrayIter(to4d(([this_batch_img])), mx.nd.array(([this_batch_lbl])), batch_size=batch_size)
            train_iter = mx.io.NDArrayIter(to4d(train_img[[ind_vec]]), mx.nd.array(train_lbl[[ind_vec]]), batch_size=batch_size)
            for batch in train_iter:
                mod.forward(batch, is_train=True)       # compute predictions
            mod.update_metric(metric, batch.label)  # accumulate prediction accuracy

            for sampleInd in np.arange(batch_size):
                softmax_vec=mod.get_outputs()[0].asnumpy()[sampleInd]
                curr_loss= -np.log(softmax_vec[np.int(batch.label[0].asnumpy()[sampleInd])])
                newPlayer.update(ind, np.exp(tree_lrn_rate*curr_loss/(prob+Eps)))
            mod.backward()                          # compute gradients
            mod.update()         # update parameters using SGD

**Comment by chao1224** [comment]
Have you tried `pred = out.get_outputs()`? This can give you the prediction for last forward() operation, and you can customize your loss function per data point.

**Comment by TalFurman** [comment]
Tnx,
It solved it.
Tal

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
